File: 3678-design-task-manager/design-task-manager.py
class TaskManager:

    def __init__(self, tasks: List[List[int]]):
        # Priority queue (min-heap), but we store (-priority, -taskId, userId)
        # Why negative? -> Python heapq is min-heap, we want max-heap
        self.tasks = [
            (-priority, -taskId, userId) for userId, taskId, priority in tasks
        ]
        heapify(self.tasks)

        # Map taskId -> priority
        self.prs = {tid: pr for _, tid, pr in tasks}
        # Map taskId -> userId
        self.usrs = {tid: uid for uid, tid, _ in tasks}

    def add(self, userId: int, taskId: int, priority: int) -> None:
        # Push task into heap
        heappush(self.tasks, (-priority, -taskId, userId))
        # Store mappings
        self.prs[taskId] = priority
        self.usrs[taskId] = userId

    def edit(self, taskId: int, newPriority: int) -> None:
        # Update if changed
        if newPriority != self.prs[taskId]:
            self.prs[taskId] = newPriority
            # Push updated task into heap (lazy deletion strategy)
            heappush(self.tasks, (-newPriority, -taskId, self.usrs[taskId]))

    def rmv(self, taskId: int) -> None:
        # Invalidate task by marking userId and priority as -1
        # (Lazy deletion, won't remove immediately from heap)
        self.prs[taskId] = -1
        self.usrs[taskId] = -1

    def execTop(self) -> int:
        if not self.tasks:
            return -1

        # Pop top element (highest priority, tie by highest taskId)
        pr, tid, uid = heappop(self.tasks)
        pr, tid = -pr, -tid

        # Lazy delete loop: discard invalid or outdated heap entries
        while pr != self.prs[tid] or uid != self.usrs[tid]:
            if not self.tasks:
                return -1
            pr, tid, uid = heappop(self.tasks)
            pr, tid = -pr, -tid

        # Remove from maps
        self.rmv(tid)

        # Return userId of executed task
        return uid
        
# Your TaskManager object will be instantiated and called as such:
# obj = TaskManager(tasks)
# obj.add(userId,taskId,priority)
# obj.edit(taskId,newPriority)
# obj.rmv(taskId)
# param_4 = obj.execTop()

File: 3678-design-task-manager/design-task-manager.go
// Task struct holds priority, taskId, and userId
type Task struct {
	priority int
	taskId   int
	userId   int
}

// MaxHeap type implements heap.Interface
type MaxHeap []Task

func (h MaxHeap) Len() int { return len(h) }

func (h MaxHeap) Less(i, j int) bool {
	// Higher priority first, if tie then higher taskId first
	if h[i].priority == h[j].priority {
		return h[i].taskId > h[j].taskId
	}

	return h[i].priority > h[j].priority
}

func (h MaxHeap) Swap(i, j int) { h[i], h[j] = h[j], h[i] }

func (h *MaxHeap) Push(x interface{}) {
	*h = append(*h, x.(Task))
}

func (h *MaxHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

// TaskManager struct
type TaskManager struct {
	h     MaxHeap
	prs   map[int]int // taskId -> priority
	usrs  map[int]int // taskId -> userId
}

// Constructor
func Constructor(tasks [][]int) TaskManager {
	h := MaxHeap{}
	prs := make(map[int]int)
	usrs := make(map[int]int)

	for _, t := range tasks {
		userId, taskId, priority := t[0], t[1], t[2]
		heap.Push(&h, Task{priority, taskId, userId})
		prs[taskId] = priority
		usrs[taskId] = userId
	}

	heap.Init(&h)

	return TaskManager{h, prs, usrs}
}

// Add task
func (this *TaskManager) Add(userId int, taskId int, priority int) {
	heap.Push(&this.h, Task{priority, taskId, userId})
	this.prs[taskId] = priority
	this.usrs[taskId] = userId
}

// Edit task priority
func (this *TaskManager) Edit(taskId int, newPriority int) {
	if this.prs[taskId] != newPriority {
		this.prs[taskId] = newPriority
		heap.Push(&this.h, Task{newPriority, taskId, this.usrs[taskId]})
	}
}

// Remove task (lazy deletion)
func (this *TaskManager) Rmv(taskId int) {
	this.prs[taskId] = -1
	this.usrs[taskId] = -1
}

// Execute the top task
func (this *TaskManager) ExecTop() int {
	for this.h.Len() > 0 {
		top := heap.Pop(&this.h).(Task)
		if this.prs[top.taskId] == top.priority && this.usrs[top.taskId] == top.userId {
			this.Rmv(top.taskId)

			return top.userId
		}
	}

	return -1
}

/**
 * Your TaskManager object will be instantiated and called as such:
 * obj := Constructor(tasks);
 * obj.Add(userId,taskId,priority);
 * obj.Edit(taskId,newPriority);
 * obj.Rmv(taskId);
 * param_4 := obj.ExecTop();
 */